raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1037,6 +1047,35 @@ private boolean handleFetchResponse(
                     logger.info("Truncated to offset {} from Fetch response from leader {}",
                         truncationOffset, quorum.leaderIdOrNil());
                 });
+            } else if (partitionResponse.snapshotId().epoch() >= 0 ||
+                       partitionResponse.snapshotId().endOffset() >= 0) {
+                // The leader is asking us to fetch a snapshot
+
+                if (partitionResponse.snapshotId().epoch() < 0) {
+                    throw new KafkaException(
```

Comment:
Hmm.. The leader has sent a bad response. I think logging an error and retrying might be better than crashing the followers.

Reply:
Yeah. I am not sure. If these exception are thrown then it means that the epoch is "valid" but the end offset is not "valid" or vice versa. I think this could happen because of either a buggy remote replica or corrupted data in the remote replica.

I guess you can argue that it is safe to keep retrying.

In the case that the `Fetch` never succeed do we want to transition to the candidate the state? If so, let me add a test that confirms that.

Reply:
Hmm.. I would rather not handle this case specially by transitioning to a candidate. It's not a crazy idea, but I think we should attempt that in a separate PR and consider errors more holistically.

I would suggest that we log an error saying that the remote replica seemed to return an invalid response and just keep fetching. Then a user can see the log message and restart the remote replica.

Reply:
> I would suggest that we log an error saying that the remote replica seemed to return an invalid response and just keep fetching. Then a user can see the log message and restart the remote replica.

Yeah. This is what I implemented and added a test for it. In other words.
1. Log an error message
2. Tell the raft client that the response was handle successfully but the fetch timer was not reset

In practice this results in the follower continuing to send `Fetch` requests. After `fetchTimeoutMs` the follower will transition to candidate as the existing client code does. See https://github.com/apache/kafka/pull/9553/files#diff-86474ad1438150630c21b29a3da2f6dd79d1357e33ac034f00e5fcef0f2e889cR350

Let me know if this is what you were thinking.